File: 1590-make-sum-divisible-by-p/1590-make-sum-divisible-by-p.py
from collections import defaultdict
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        if sum(nums)%p==0 : return 0
        n = len(nums)

        s = sum(nums)
        target = s%p

        mp = defaultdict(int)

        mp[0] = -1

        cur = 0
        ans = n

        for i in range(n):
            cur = (cur + nums[i]) % p

            needed = (cur - target) % p

            if needed in mp:
                ans = min(ans, i - mp[needed])

            mp[cur] = i

        return ans if ans < n else -1
        

File: 1590-make-sum-divisible-by-p/1590-make-sum-divisible-by-p.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        long long s = accumulate(begin(nums) , end(nums) , 0LL);

        if (s % p == 0) return 0;

        int target = s % p;

        unordered_map<int , int> mp;

        mp[0] = -1;

        int cur = 0;
        int ans = nums.size();

        for (int i = 0 ; i<nums.size() ; i++){
            cur = (cur + nums[i]) % p;

            int needed = (cur - target + p) % p;

            if (mp.count(needed)){
                ans = min(ans , i-mp[needed]);
            }

            mp[cur] = i;

        }

        return (ans<nums.size()) ? ans : -1 ;

    }
};